perf(parser): use visitor instead of JSON.parse reviver

ArnaudBarre opened the pull request:
Fixes #10783

This requires having a list of ~135 nodes with their non primitive keys. 
Technically this list could be generated alongside the raw deserializer but I don't know this code enough to do it.
I've chosen to inline visitor keys to avoid adding [3 dependencies](https://pkg-size.dev/@typescript-eslint%2Fvisitor-keys) for that.
I didn't find how the main LICENCE was injected during the publish so I couldn't append the bundled LICENSES and choose to use `@licence` comments with links

ArnaudBarre committed:
perf(parser): use visitor instead of JSON.parse reviver

overlookmotel commented:
I've squashed to 1 commit and rebased on main, just to fix merge conflicts and make it easier to review.

I'll try and figure out cause of the test failures.

overlookmotel committed:
Remove licenses

We import `@typescript-eslint/visitor-keys` as a dependency, rather than copying it's code. So no need for license attribution.

overlookmotel committed:
Refactor generator

overlookmotel committed:
Standardize generated file header

overlookmotel committed:
Allow generate script to be run from anywhere

overlookmotel committed:
Check in generated code to git

overlookmotel committed:
refactor: add blocks to `for` loops

overlookmotel committed:
Rename vars

overlookmotel committed:
Add visitor keys for `ParenthesizedExpression` and `TSParenthesizedType`

overlookmotel committed:
Fix `wrap.mjs`

overlookmotel commented:
I found the problem. Our AST has 2 node types which TS-ESLint doesn't:

* `ParenthesizedExpression`
* `TSParenthesizedType`

Have pushed a commit to fix, and also done a little refactoring to (in my opinion) make the code a bit clearer.

ArnaudBarre commented:
Thanks for finding that! Changes looks good to me!
